### ISSUE TITLE
ReadCoreVersion fails on OS X when compiling with FPC 3.0.0

### ISSUE THREAD

**Issue by basisbit** [SaschaWillems/dglOpenGL#4]
We, the UltraStar Deluxe team, are currently preparing a new release and am now trying to get the SDL2 based game to work on OS X. Sadly, after switching to dglOpenGL, executing the game fails on Mac OS X el capitain in dglOpelGL.pas procedure ReadCoreVersion line 19615 `AnsiBuffer := glGetString(GL_VERSION);`

Compiling works without any problem and dglOpenGL works without any problem on Linux and Windows based systems.
Any idea what could cause this issue?

We call the function from https://github.com/UltraStar-Deluxe/USDX/blob/master/src/base/UGraphic.pas#L681
Thanks in advance!

**Comment by basisbit** [comment]
at least partially fixed with pull request #5

**Comment by SaschaWillems** [comment]
Sry for the late reply, but I don't have a Mac OSX system to test with me right now, so unless you find a solution I'll have to find a Mac somewhere. Basically this should work, as some of my old FPC/OpenGL suff works on Mac OSX.

**Comment by neurolabusc** [comment]
@basisbit this command works in my [PlyView](https://github.com/neurolabusc/plyview/blob/master/unit1.pas) project. The glGetString(GL_VERSION) function inserts the version into the form's caption at launch. For my project, you need to use the Cocoa widgetset (Project/ProjectOptions/Additions&Overrides, change "Default" to "MacOS"). By default my project uses the corearb library, but you can switch this to dgl by uncommenting `//` from the line `// {$DEFINE DGL}`. This project requires SVN > 55355 for Retina support, but you can run on older versions by commenting out the line `{$DEFINE RETINA}`

**Comment by neurolabusc** [comment]
@basisbit - if you look at my [plyview](https://github.com/neurolabusc/plyview/) code you will see how a handful of IFDEF commands can switch between the GL/GLEXT, dglOpenGL and glcorearb libraries for OpenGL support. For your purposes you may find using the GL/GLEXT libraries the best match, and use `Load_GL_version_2_1()` to load legacy OpenGL. The 'legacy' folder of [OpenGLCoreTutorials](https://github.com/neurolabusc/OpenGLCoreTutorials) provides several examples using GL/GLEXT for OpenGL 2.1 projects that compile on Linux, Windows and macOS.

Looking at your code, you use legacy OpenGL immediate-mode commands: glMatrixMode, GL_QUADS, glBegin. Be aware that these do not exist in Core OpenGL's retained mode. For cross-platform projects that target macOS you **MUST** target either 
1. OpenGL 2.1 or earlier 
2. OpenGL 3.2 or later **CORE** profile.

 On Windows and Linux versions of OpenGL after 2.1 support the compatibility profile, which allows you to mix and match legacy immediate mode calls with core retained mode calls. With macOS, if you request a version of OpenGL greater than 2.1, you will get the **CORE** profile which removes all legacy functions. Calling a legacy command in Core mode will either do nothing or crash the program. To reiterate, if you run OpenGL 3.3 on Linux or Windows it is a superset of all prior OpenGL commands. On macOS if you run OpenGL 3.3 you are in the core profile and have no access to legacy features. This is not a limitation of dglOpenGL: it is a "feature" of the macOS. The core mode forces users to write their own shaders and efficiently retain data on the GPU. Handy but computationally inefficient (in particular with respect to CPU-GPU memory bandwidth) features were eliminated from the Core mode. 

Here are the pros and cons of each library:

1. dglOpenGL benefits: same code for Delphi and Lazarus, automatically detects latest available OpenGL features. Cons: on macOS it loads the latest OpenGL core mode, and any calls to legacy OpenGL can cause a crash.
2. GL/GLEXT benefits: included and well-supported with Lazarus. Cons: will crash if you request OpenGL version > 2.1 for macOS. Therefore, use `Load_GL_version_2_1()` if you wish to target macOS, Linux and Windows.
3. glcorearb benefits: only exposes CORE features of OpenGL. Code will not compile if a user attempts to use a feature not provided in core mode (e.g. GL_QUADS is not defined, so program does not compile rather than compiling and crashing on MacOS). Allows users to write OpenGL 3.3 code that will work on MacOS, Windows and Linux. Cons: does not support any legacy OpenGL features.

**Comment by basisbit** [comment]
tldr: thanks for your analyses, but I don't care about mac OS support any more and found a workaround for this issue -> thus closed the issue.
works on Windows and Linux. Also works mostly on mac OS. I am not gonna touch the old freepascal code mess of UltraStar Deluxe more then necessary. And I don't care about mac OS support. 
If you think anything should be fixed / changed, feel free to create a pull request and / or join the dev team. :)

**Comment by basisbit** [comment]
on a side note, why do so many devs on mac OS implement "defineRetina" switches? This is rediculous.  Code should be resolution-independent, and even if it is not, then that should be named defineHighRes or something like that. "Retina" is just a marketing word in the apple world.

**Comment by neurolabusc** [comment]
It does appear that dglOpenGL can load a OpenGL 2.1 context and call `glGetString(GL_VERSION)` on macOS. For example, load the 
[basicLegacy project](https://github.com/neurolabusc/OpenGLCoreTutorials) and uncomment `//{$DEFINE DGL}` - note it reports version 2.1 in the titlebar when the program launches. 

I suspect that the difference is the way we create our context. In my Plyview project I explicitly request OpenGL 3.3 and in the basicLegacy I explicitly set the context to 2.1. However, I suspect you are simply requesting the latest version in your SDL call, so you are getting a core context even though your code is designed for a legacy context. Can you modify SDL_GL_CreateContext to set up a 2.1 context?
```
  glcontext := SDL_GL_CreateContext(Screen);
  InitOpenGL();
```
By the way, this is a very nice project.

**Comment by neurolabusc** [comment]
Actually, it is not the developers but rather how [Apple](https://developer.apple.com/library/content/documentation/GraphicsAnimation/Conceptual/HighResolutionOSX/CapturingScreenContents/CapturingScreenContents.html) defines their [OpenGL 
 API via wantsBestResolutionOpenGLSurface](https://developer.apple.com/documentation/appkit/nsview/1414938-wantsbestresolutionopenglsurface). For the rest of the mac interface, you set up retina mode using when the application starts up uing the Info.plist. However, for OpenGL you can turn retina on and off dynamically. Retina is a lot slower though nicer looking, so it is often useful to render things at the lower resolution.

-c

On Aug 7, 2017, at 12:20 PM, basisbit <notifications@github.com<mailto:notifications@github.com>> wrote:


on a side note, why do so many devs on mac OS implement "defineRetina" switches? This is rediculous. Code should be resolution-independent, and even if it is not, then that should be named defineHighRes or something like that. "Retina" is just a marketing word in the apple world.
